README.txt
vs.event
========

``vs.event`` provides an extended event functionality for Plone

Features
========

- recurring events
- a new calendar widget 
- real support for all-day events
- extended support for attendees and attachments (UI and iCal level)
- working iCal/vCal export for all-day events
- full integration with Plone4Artists calendar (must be installed seperately)
- support for supplementary-events (one master event and several depending events)

Supplementary events
====================

There are case where some particular event in reality consists of several date.
E.g. for conference  you need an event for the conference date itself. However
you often have additional supplementary events like a deadline for a
call-for-papers or a registration deadline. With ``vs.event`` you create an
event for the conference itself and so-called ``supplementary events`` for the
call-for-papers and the registration deadline. The supplementary dates are
events themselves but linked with the primary event (using the Archetypes
references). So with ``vs.event`` you will see all supplementary events for the
main event or jump directly from a supplementary event to its main event.  This
feature is optional and can be disabled through the ZMI in portal_calendar ->
vs_event_supplementary_events.

Integration with Plone4Artists calendar
=======================================

``vs.event`` integrates smoothly with ``p4a.plonecalendar``. A new view
``icalendar_export`` is registered for p4a.plonecalendar enabled folders.  It
will export all ``vs.event`` related events in iCal. This means you can
subscribe with iCal to a calendar taking full advantage of the `vs.event``
functionality (including support for attachments and attendees on the iCal
level).

Installation
============

Add ``vs.event`` to the ``eggs`` and ``zcml`` options of your buildout.cfg.
Create a new Plone site using the ``vs.event`` profile or install ``vs.event``
through the quick installer of Plone.

Known bugs and limitations
==========================

- the localization of the date picker widget support only 'en' and 'de'
  so far. The date picker widget will use/present the US date format for
  languages other than German.

TODO
====

- more i18n needed

License
=======

``vs.event`` is published under the GNU Public License 2.

Parts of the code of ``vs.event`` (iCal implementation) are based on work
in ATContentTypes.

Authors
=======

- Andreas Jung
- Veit Schiele
- Anne Walther

Repository
==========

http://svn.plone.org/svn/collective/vs.event

Contact
=======

Send mail to vs.event@veit-schiele.de



docs/HISTORY.txt
Changelog
=========

0.9.0 - (Unreleased)
--------------------

- Plone 4.3 compatibility
  [ale-rt]

0.3.2 - (Unreleased)
--------------------
- add french translation [toutpt]

0.3.1 - (2011-12-05)
--------------------

0.3.0 - (2011-12-05)
---------------------
- vs_event_view - fixes for plone4
- compatibility fixes
- *requires* Plone 4.X

0.2.19 - (2009/11/10)
---------------------

* i18n fixes

0.2.18 - (2009/08/31)
---------------------

* fixed import bug with dt2DT()

0.2.17 - (2009/08/09)
---------------------

* fix for recurrence and sub events within event_util.getICal()

0.2.16 - (2009/07/22)
---------------------

* some i18n fixes
* hidding subscription URL in view mode since this is information is
  redundant


0.2.15 - (2009/07/08)
---------------------

* fixed import issues in browser view code (reported by Daniel Widerin)


0.2.14 - (2009/07/08)
---------------------

* 'until' field appeared within the wrong schemata


0.2.13 - (2009/07/08)
---------------------

* fix in JS profile

0.2.12 - (2009/07/05)
---------------------

* removed p4a.common dependency

0.2.11 - (2009/07/05)
---------------------

* removed p4a.ploneevent dependency (requires p4a.plonecalendar 2.0b5+)

0.2.10 - (2009/06/24)
---------------------

* post_validator() now takes the value of 'useEndDate' into account

0.2.9 - (2009/06/21)
--------------------

* fixed racing condition in JS code

0.2.8 - (2009/06/21)
--------------------

* updated locales
* using icalendar_export_event view as a replacement for ics_view
* ical export for an event with supplementary events now exports
  all related events instead of having to subscribe each individual
  event

0.2.7 - (2009/06/19)
--------------------

* fixed p4a.common dependency

0.2.6 - (2009/06/18)
--------------------

* minor fixes and UI tweaks

0.2.5 - (2009/06/18)
--------------------

* timezone fix in iCal export: whole day events used
  a shortened date string based on UTC - but we need
  a date string based on the local zone.


0.2.4 - (2009/06/17)
--------------------

* minor fix in icalendar_export functionality

0.2.3 - (2009/06/17)
--------------------

* display list of attachments in 'view' mode
* fixed tests a bit
* updated documentation


0.2.2 - (2009/06/16)
--------------------

* added new field for attachments (UI + iCal)
* added new field for attendees (UI + iCal)

0.2.1 - (2009/06/13)
--------------------

* test suite
* re-worked date format string handling
* minor fixes

0.2.0 - (2009/06/12)
--------------------

* fixes and cleanup

0.1.3 - (2009/06/12)
--------------------

* fixes


0.1.1 - (2009/06/12)
--------------------

* various fixes related to the localization of the datepicker widget


0.1.0 - (2009/06/09)
--------------------

* Initial release


vs/event/content/event.py
# -*- coding: utf-8 -*-
# $Id: event.py 388 2009-06-13 09:32:31Z ajung $
from AccessControl import ClassSecurityInfo
from DateTime.DateTime import DateTime
from Products.ATContentTypes.content.event import ATEvent
from Products.ATContentTypes.content.schemata import finalizeATCTSchema
from Products.CMFCore.permissions import View
from Products.DataGridField import DataGridField
from Products.DataGridField import DataGridWidget
from Products.DataGridField.CheckboxColumn import CheckboxColumn
from Products.DataGridField.Column import Column
from Products.DataGridField.SelectColumn import SelectColumn
from archetypes.referencebrowserwidget.widget import ReferenceBrowserWidget
from collective.calendarwidget.widget import CalendarWidget
from dateutil.rrule import YEARLY
from dateutil.rrule import MONTHLY
from dateutil.rrule import WEEKLY
from dateutil.rrule import DAILY
from vs.event import MessageFactory as _
from vs.event.config import PROJECTNAME
from vs.event.interfaces import IVSEvent
from vs.event.interfaces import IVSSubEvent
from zope.interface import implements
from . import event_util

try:
    from Products.LinguaPlone import public as atapi
except ImportError:
    from Products.Archetypes import atapi


from logging import getLogger
log = getLogger(">> event : ")

VSEventSchema = atapi.Schema((

    atapi.ReferenceField(
        name='subEvents',
        allowed_types=('VSSubEvent',),
        multiValued=True,
        relationship='isSubEvent',
        widget=ReferenceBrowserWidget(
            visible={'view': 'invisible', 'edit': 'invisible'},
        ),
    ),
    atapi.LinesField(
        name='weekdays',
        schemata='recurrence',
        vocabulary=atapi.DisplayList((
            ('0', _(u'vs_label_mo', 'Monday')),
            ('1', _(u'vs_label_di', 'Tuesday')),
            ('2', _(u'vs_label_mi', 'Wednesday')),
            ('3', _(u'vs_label_do', 'Thursday')),
            ('4', _(u'vs_label_fr', 'Friday')),
            ('5', _(u'vs_label_sa', 'Saturday')),
            ('6', _(u'vs_label_so', 'Sunday'))
        )),
        widget=atapi.MultiSelectionWidget(
            format="checkbox",
            label=_(u'vs_event_label_weekdays', 'Weekdays'),
            description=_(u'vs_event_help_weekdays', 'Select weekdays'),
        ),
    ),
    atapi.StringField(
        name='bysetpos',
        schemata='recurrence',
        validators=('isLineOfInts',),
        widget=atapi.StringWidget(
            label=_(u'vs_event_label_bysetpos', 'Occurrence'),
            description=_(
                u'vs_event_help_bysetpos', 'Comma separated list of '
                'numbers. If this event is on first and third '
                'Monday of the month, enter "1,3" and select '
                'appropriate day of the week above. For the last '
                'day of the month, enter "-1", select Monday '
                'through Friday above and monthly recurrence.'
            ),
        ),
    ),
    atapi.LinesField(
        name='exceptions',
        schemata='recurrence',
        validators=('linesOfDates',),
        widget=atapi.LinesWidget(
            label=_(u'vs_event_label_exceptions', 'Exceptions'),
            description=_(
                u'vs_event_help_exceptions',
                (
                    'Please enter exceptions to recurrence. '
                    'One date per line in format YYYY-MM-DD')
            ),
        ),
    ),

    atapi.IntegerField(
        'frequency',
        schemata="recurrence",
        required=True,
        i18n_domain = "vs.event",
        vocabulary={
            str(-1): _(u'Does not repeat'),
            str(YEARLY): _(u'Yearly'),
            str(MONTHLY): _(u'Monthly'),
            str(WEEKLY): _(u'Weekly'),
            str(DAILY): _(u'Daily'),
        }.items(),
        default=-1,
        widget=atapi.SelectionWidget(
            label=_(u"vs_event_label_frequency", 'Frequency'),
        )
    ),
    atapi.IntegerField(
        'interval',
        schemata="recurrence",
        required=True,
        default=1,
        widget=atapi.IntegerWidget(
            label=_(u"vs_event_label_interval", 'Interval'),
        )
    ),
    atapi.DateTimeField(
        'until',
        schemata="recurrence",
        widget=CalendarWidget(
            label=_(u"vs_event_label_repeat_until", "Repeat until"),
            description=_(
                u"vs_event_description_repeat__until",
                "Event repeats until this date."),
            with_time=1)
    ),
    atapi.IntegerField(
        'count',
        schemata="recurrence",
        widget=atapi.IntegerWidget(
            label=_(u"vs_event_label_count", 'Count'),
            description=_(
                u"vs_event_description_count",
                "Maxinum number of times the event repeats"),
        )
    ),

))

VSEventSchema = VSEventSchema + ATEvent.schema.copy()
finalizeATCTSchema(VSEventSchema, folderish=False, moveDiscussion=True)


def modifyEventSchema(schema):
    schema.addField(atapi.BooleanField(
        'wholeDay',
        widget=atapi.BooleanWidget(
            label=_(u'label_whole_day_event', 'Whole day event'),
        ),
    ))
    schema.addField(atapi.BooleanField(
        'useEndDate',
        default=True,
        widget=atapi.BooleanWidget(
            label=_(u'label_use_end_date', 'Use end date?'),
        ),
    ))
    schema.moveField('wholeDay', before='startDate')
    schema.moveField('useEndDate', after='wholeDay')
    schema['startDate'].widget = CalendarWidget(
        label=_(u"label_event_start", "Event Starts"),
        with_time=1,
        with_popup=1,
        js_shortcuts=0,
        ignore_unset_time=1)

    schema['endDate'].widget = CalendarWidget(
        label=_(u"label_event_end", "Event Ends"),
        with_time=1,
        with_popup=1,
        js_shortcuts=0,
        ignore_unset_time=1)

    del schema['until']
    schema.addField(
        atapi.DateTimeField(
            'until',
            schemata='recurrence',
            widget=CalendarWidget(
                description=_(
                    u"vs_repeat_events_until_date",
                    "Event repeats until this date"),
                label=_(
                    u"vs_label_repeat_until",
                    "Repeat until"),
                with_time=1)))

    schema.addField(
        atapi.ReferenceField(
            name='attachments',
            allowed_types=('Link', 'File'),
            multiValued=1,
            relationship='aAttachment',
            widget=ReferenceBrowserWidget(
                show_review_state=1,
                allow_sorting=1,
                force_close_on_insert=1,
                base_query={'Type': ['Link', 'File']},
                label=_(u'vs_event_label_attachments', 'Attachments'),
            ),
        ))

    del schema['attendees']
    schema.addField(
        DataGridField(
            name="attendees",
            columns=('name', 'mail', 'role', 'show'),
            schemata='attendees',
            widget=DataGridWidget(
                label=_(
                    u'vs_event_label_roleAttendees',
                    "Attendees and roles"
                ),
                columns={
                    'name': Column(_(u'vs_event_label_nameColumn', "Name")),
                    'mail': Column(_(u'vs_event_label_mailColumn', "e-Mail")),
                    'role': SelectColumn(
                        _(u'vs_event_label_roleColumn', "Role"),
                        vocabulary='getAttendeeRoles'
                    ),
                    'show': CheckboxColumn(
                        _(u'vs_event_label_showColumn', "Show")
                    ),
                }),
        ))
    schema.moveField('attendees', after='count')
    return schema


class VSEvent(ATEvent):

    """ vs.event """

    security = ClassSecurityInfo()
    implements(IVSEvent)
    meta_type = 'VSEvent'
    _at_rename_after_creation = True
    schema = modifyEventSchema(VSEventSchema)

    security.declareProtected(View, 'getICal')

    def getICal(self):
        """get iCal data """
        return event_util.getICal(self)

    security.declareProtected(View, 'getVCal')

    def getVCal(self):
        """get VCal data """
        return event_util.getVCal(self)

    def at_post_edit_script(self):
        """ Ensure that for single-day dates without an end date
            the end date is equal to the start date.
        """

        self.setExcludeFromNav(True)

        if not self.getUseEndDate():
            self.setEndDate(self.start())

        if self.getWholeDay():
            self.setStartDate(
                DateTime(self.start().strftime('%Y/%m/%d 00:00:00'))
            )
            self.setEndDate(
                DateTime(self.end().strftime('%Y/%m/%d 23:59:00'))
            )

    def getAttendeeRoles(self):
        """ """
        return atapi.DisplayList(
            (
                (
                    'chair',
                    _(u'vs_event_label_chair', "Chair")
                ),
                (
                    'observer',
                    _(u'vs_event_label_observer', "Observer")
                ),
                (
                    'participant',
                    _(u'vs_event_label_participant', "Participant")
                ),
                (
                    'opt_participant',
                    _(
                        u'vs_event_label_opt_participant',
                        "Optional participant"
                    )
                ),
            )
        )

    security.declareProtected(View, 'post_validate')

    def post_validate(self, REQUEST=None, errors=None):
        """ Trigger original ATEvent.post_validate_method() for
            dates having an end date.
        """
        if REQUEST.get('useEndDate', True):
            return super(VSEvent, self).post_validate(
                REQUEST=REQUEST,
                errors=errors
            )

atapi.registerType(VSEvent, PROJECTNAME)


def modifySubEventSchema(schema):
    # remove unwanted fields for subevents

    # PLONE 4 seems not to have a eventType field anymore, so it must be
    # excluded here for compatibility
    for id in (
            'attendees',
            'contactName',
            'contactEmail',
            'contactPhone',
            'eventUrl'):
        schema[id].widget.visible = False
    for field in schema.fields():
        if field.schemata in (
                'dates',
                'categorization',
                'ownership',
                'settings'):
            field.widget.visible = False

    schema.addField(
        atapi.BooleanField(
            'wholeDay',
            default=False,
            widget=atapi.BooleanWidget(
                label=_(u'label_whole_day_event', 'Whole day event'),
            ),
        )
    )
    schema.addField(
        atapi.BooleanField(
            'useEndDate',
            default=True,
            widget=atapi.BooleanWidget(
                label=_(u'label_use_end_date', 'Use end date?'),
            ),
        )
    )
    schema.moveField('wholeDay', before='startDate')
    schema.moveField('useEndDate', after='wholeDay')

    schema['startDate'].widget = CalendarWidget(
        label=_(u"label_event_start", "Event Starts"),
        with_time=1,
        with_popup=1,
        js_shortcuts=0,
        ignore_unset_time=1
    )

    schema['endDate'].widget = CalendarWidget(
        label=_(u"label_event_end", "Event Ends"),
        with_time=1,
        with_popup=1,
        js_shortcuts=0,
        ignore_unset_time=1
    )

    return schema

VSSubEventSchema = ATEvent.schema.copy()
finalizeATCTSchema(VSSubEventSchema, folderish=False, moveDiscussion=True)


class VSSubEvent(VSEvent):

    """ vs.event """

    implements(IVSSubEvent)
    meta_type = 'VSSubEvent'
    schema = modifySubEventSchema(VSSubEventSchema)

atapi.registerType(VSSubEvent, PROJECTNAME)
